[COMPRESS-695] Configure ZipFile with a custom Zstd input stream implementation

mehmet-karaman opened the pull request:
Adjusted the ZipFile and the Builder. Added a Test for checking both.

Jira: https://issues.apache.org/jira/browse/COMPRESS-695

mehmet-karaman commented:
Build fails because of JavaDoc warning .. But I've not touched the Attribute.java.. Is this normal?

[WARNING] Javadoc Warnings
[WARNING] /home/mkaraman/git/commons-compress/src/main/java/org/apache/commons/compress/harmony/unpack200/bytecode/Attribute.java:129: warning: no description for @throws
[WARNING] * @throws IOException
[WARNING] ^
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:57 min
[INFO] Finished at: 2025-03-21T13:46:13+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:3.11.2:javadoc (default-cli) on project commons-compress: An error has occurred in Javadoc report generation: Project contains Javadoc Warnings -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:

garydgregory commented on src/main/java/org/apache/commons/compress/archivers/zip/ZipFile.java:
```diff
@@ -167,7 +168,26 @@ public ZipFile get() throws IOException {
+         * @return {@code this} instance
+         * @since 1.28.0
+         */
+        public Builder setZstdInputStreamCreator(IZstdInputStreamCreator zstdInpStreamCreator) {
```
The usual naming convention for this pattern is "factory", not a "creator". See above RE using a `Function`.

reset to `ZstdCompressorInputStream::new` if null.

mehmet-karaman replied:
done.

garydgregory commented on src/main/java/org/apache/commons/compress/archivers/zip/ZipFile.java:
```diff
@@ -138,6 +138,7 @@ public static class Builder extends AbstractStreamBuilder<ZipFile, Builder> {
         private boolean useUnicodeExtraFields = true;
         private boolean ignoreLocalFileHeader;
         private long maxNumberOfDisks = 1;
+        private IZstdInputStreamCreator zstdInpStreamCreator;
```
There is no need for a custom type or odd names parts like "Inp":
```
private IOFunction<InputStream,InputStream> zstdInputStream = ZstdCompressorInputStream::new;
```

mehmet-karaman replied:
done.

garydgregory commented on src/main/java/org/apache/commons/compress/archivers/zip/ZipFile.java:
```diff
@@ -167,7 +168,26 @@ public ZipFile get() throws IOException {
+            return new ZipFile(actualChannel, actualDescription, getCharset(), useUnicodeExtraFields, closeOnError, ignoreLocalFileHeader) {
+                @Override
+                protected InputStream createZstdInputStream(InputStream is) throws IOException {
+                    if (zstdInpStreamCreator != null) {
```
Do the null check in the setter and reset to `ZstdCompressorInputStream::new` if null.

mehmet-karaman replied:
done.

garydgregory commented on src/test/java/org/apache/commons/compress/archivers/zip/ZipFileTest.java:
```diff
@@ -66,8 +66,53 @@
 public class ZipFileTest extends AbstractTest {
 
+    private final class ZstdInputStreamCreator implements IZstdInputStreamCreator {
+        private boolean isUsed;
```
Use the same naming convention as other test fixture: `isUsed` -> `used`.

mehmet-karaman replied:
I've changed that to instance check.. so this is also obsolete..

garydgregory commented on src/test/java/org/apache/commons/compress/archivers/zip/ZipFileTest.java:
```diff
@@ -66,8 +66,53 @@
+
 public class ZipFileTest extends AbstractTest {
 
+    private final class ZstdInputStreamCreator implements IZstdInputStreamCreator {
```
Add a Javadoc comment as to why we need a test class.

mehmet-karaman replied:
obsolete, due we switched to IOFunction.

garydgregory commented:
> Build fails because of JavaDoc warning .. But I've not touched the Attribute.java.. Is this normal?
> 
> [WARNING] Javadoc Warnings [WARNING] /home/mkaraman/git/commons-compress/src/main/java/org/apache/commons/compress/harmony/unpack200/bytecode/Attribute.java:129: warning: no description for @throws [WARNING] * @throws IOException [WARNING] ^ [INFO] ------------------------------------------------------------------------ [INFO] BUILD FAILURE [INFO] ------------------------------------------------------------------------ [INFO] Total time: 01:57 min [INFO] Finished at: 2025-03-21T13:46:13+01:00 [INFO] ------------------------------------------------------------------------ [ERROR] Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:3.11.2:javadoc (default-cli) on project commons-compress: An error has occurred in Javadoc report generation: Project contains Javadoc Warnings -> [Help 1] [ERROR] [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch. [ERROR] Re-run Maven using the -X switch to enable full debug logging. [ERROR] [ERROR] For more information about the errors and possible solutions, please read the following articles:

The GH CI build disables Javadoc linting, see the maven.yml file.

mehmet-karaman commented on src/main/java/org/apache/commons/compress/archivers/zip/IZstdInputStreamCreator.java:
```diff
@@ -0,0 +1,35 @@
+ * This interface can be implemented to support another Zstd compression library.
+ * Due the {@link ZipFile}s constructor is deprecated, this Interface can be used to pass it to the Builder.
+ */
+public interface IZstdInputStreamCreator {
```
Thanks for this hint, Removed that interface and used the IOFunction. :)

mehmet-karaman commented on src/main/java/org/apache/commons/compress/archivers/zip/IZstdInputStreamCreator.java:
```diff
@@ -0,0 +1,35 @@
+public interface IZstdInputStreamCreator {
+
+    /**
+     * @param is the input stream which should be used for compression.
```
Not necessary anymore as its replaced by IOFunction.

mehmet-karaman commented:
Thank you for the Review.. I have tried to use the IOFunction.. It turns out that the commons-io has many helpful utilities :)

garydgregory commented:
Hello @mehmet-karaman 

Please don't push broken builds. Run `mvn` locally (by itself) first and make sure the build passes.

mehmet-karaman commented:
Hello @garydgregory ,

sorry.. I've forgotten that.. I've fixed the checkstyle glitch (and also fixed the Javadoc problem of the Attributes class which failed again). Local mvn was successful.

garydgregory commented on src/main/java/org/apache/commons/compress/archivers/zip/ZipFile.java:
```diff
@@ -167,7 +169,26 @@ public ZipFile get() throws IOException {
+        }
+
+        /**
+         * @param zstdInpStreamFactory the IOFunction which gives the ability to return a different InputStream for Zstd compression.
```
This method is missing a description.

mehmet-karaman replied:
done.

garydgregory commented on src/main/java/org/apache/commons/compress/archivers/zip/ZipFile.java:
```diff
@@ -167,7 +169,26 @@ public ZipFile get() throws IOException {
+            if (zstdInpStreamFactory == null) {
+                this.zstdInputStream = ZstdCompressorInputStream::new;
+            } else {
+                this.zstdInputStream = zstdInpStreamFactory;
```
Simplify: Use a ternary expression.

mehmet-karaman replied:
done.

garydgregory commented on src/test/java/org/apache/commons/compress/archivers/zip/ZipFileTest.java:
```diff
@@ -66,8 +66,36 @@
+    private final class AirliftZstdZipFile extends ZipFile {
+        private boolean used;
+
+        private AirliftZstdZipFile(File file) throws IOException {
```
Parameter: Use `final`.

mehmet-karaman replied:
done

garydgregory commented on src/test/java/org/apache/commons/compress/archivers/zip/ZipFileTest.java:
```diff
@@ -66,8 +66,36 @@
+            super(file);
+        }
+
+        protected InputStream createZstdInputStream(InputStream is) throws IOException {
```
Parameter: Use `final`.

mehmet-karaman replied:
done

garydgregory commented on src/test/java/org/apache/commons/compress/archivers/zip/ZipFileTest.java:
```diff
@@ -66,8 +66,36 @@
+            return new ZstdInputStream(is) {
+
+                @Override
+                public int read(byte[] outputBuffer, int outputOffset, int outputLength) throws IOException {
```
Parameters: Use final.

mehmet-karaman replied:
done

garydgregory commented on src/main/java/org/apache/commons/compress/archivers/zip/ZipFile.java:
```diff
@@ -167,7 +169,26 @@ public ZipFile get() throws IOException {
             final boolean closeOnError = seekableByteChannel != null;
-            return new ZipFile(actualChannel, actualDescription, getCharset(), useUnicodeExtraFields, closeOnError, ignoreLocalFileHeader);
+            return new ZipFile(actualChannel, actualDescription, getCharset(), useUnicodeExtraFields, closeOnError, ignoreLocalFileHeader) {
+                @Override
```
Hello @mehmet-karaman 

Thank you for updating this pull request.

The point of parameterizing the builder is to pass the `zstdInputStream` to the `ZipFile` constructor which can then be recoded to apply the function in `createZstdInputStream()`. Change the `ZipFile`'s `createZstdInputStream()` implementation to use `zstdInputStream`, then it's the builder that has the only hard-coded reference to `ZstdCompressorInputStream`. The `ZipFile` constructor to edit is private, so we are keeping binary compatibility.

mehmet-karaman replied:
There is still a possiblity that some implementations are using the old deprecated constructors, so I had to initialize the ZipFiles zstdInputStream field too. So the ZipFile class and the Builder need the hardcoded reference.

mehmet-karaman replied:
Could you re-check that please? Maybe be I didn't understand what you mean with "only hard-coded" reference..

garydgregory reviewed (changes requested):
Hello @mehmet-karaman

Thank you for updating this pull request.

Please see my comments.

garydgregory commented on src/main/java/org/apache/commons/compress/archivers/zip/ZipFile.java:
```diff
@@ -994,6 +1015,14 @@ public ZipFile(final String name, final String encoding) throws IOException {
         this(new File(name).toPath(), encoding, true);
     }
 
+    private ZipFile(SeekableByteChannel actualChannel, String actualDescription, Charset charset,
```
You don't need this, the constructor `ZipFile.ZipFile(SeekableByteChannel, String, Charset, boolean, boolean, boolean)` is already `private`, so you can edit it.

mehmet-karaman replied:
good point.. now the field can be also final :)

mehmet-karaman commented:
Hi @garydgregory,

I saw that you have added a new static field which, which causes a problem. In ZipFile line 96, which makes the com.github.luben.zstd.ZstdInputStream necessary.. Now the maven dependency isn't optional anymore.

`Caused by: java.lang.ExceptionInInitializerError: Exception java.lang.NoClassDefFoundError: com/github/luben/zstd/ZstdInputStream [in thread "main"]
	at org.apache.commons.compress.archivers.zip.ZipFile.<clinit>(ZipFile.java:96)
	at org.apache.commons.compress.archivers.zip.ZipArchiveInputStream.getNextZipEntry(ZipArchiveInputStream.java:702)
	at org.apache.commons.compress.archivers.zip.ZipArchiveInputStream.getNextEntry(ZipArchiveInputStream.java:645)
	at zipeditor.model.ZipModel.detectType(ZipModel.java:119)`

Is it maybe better to extract this to a utility class?

garydgregory commented:
Hi @mehmet-karaman 

I'm not sure inlining the constant will matter because it is referred from the builder and a constructor.

Please try changes locally and see if you can improve your use case.
